source/extended-json.rst

```diff
@@ -0,0 +1,618 @@
+as having significance. For example, the document ``{"$code": "function(){}",
+"$scope": {}}`` must be considered identical to ``{"$scope": {}, "$code":
+"function(){}"}``.
+
+.. _JSON specification: https://tools.ietf.org/html/rfc7159
+
+Generators
+----------
+
+An extended JSON generator (hereafter just "generator") produces strings in
+extended JSON. A generator MUST be able to produce strings that conform strictly
+to the canonical extended JSON format. Such generators SHOULD produce such
+strings by default. If a generator does not produce strings in the canonical
+format by default, then that generator MUST be configurable to use the canonical
+format. A generator MAY be capable of exporting strings that adhere to other
+formats, such as historical extended JSON formats emitted by old versions of
+mongoexport. A generator MUST support at least 100 levels of nesting in a BSON
+document.
+
+Transforming BSON
+.................
+
+When transforming a BSON value into extended JSON text, a generator MUST use the
+corresponding format for the BSON type given in the `Canonical Format
+Table`_. When transforming a BSON document into extended JSON text, a generator
+SHOULD emit the JSON keys and values in the same order as given in the BSON
+document.
+
+Transforming Language-Specific Data Structures
+..............................................
+
+When transforming a language-specific data structure, such as one emitted by a
+BSON parser, into extended JSON, a generator MUST emit the extended JSON form
+for the BSON type corresponding to the type of the data structure. For example,
+a Python ``datetime`` object must be represented the same as a BSON datetime
+type.
+
+Extended JSON Format
+--------------------
+
+The extended JSON grammar extends the JSON grammar as defined in `section 2`_ of
+the `JSON specification`_ by augmenting the possible JSON values as defined in
+`Section 3`_. An extended JSON value MUST conform to one of the formats given in
+the second column of the following table:
+
+.. _section 2: https://tools.ietf.org/html/rfc7159#section-2
+.. _section 3: https://tools.ietf.org/html/rfc7159#section-3
+
+.. _Canonical Format Table:
+
+Canonical Format Table
+----------------------
+
+Notes
+.....
+
+* Keys within extended JSON values SHOULD be emitted in the order described. All
+  keys MUST be emitted, and any other elements MUST NOT be emitted.
+* Types in *italics* are defined elsewhere in the table or in the `JSON
+  specification`_.
+* Types in <angle brackets> detail the contents of a value.
+
++--------------------+----------------------------------------------------------+
+|**BSON 1.1 Type or  |**Extended JSON Format**                                  |
+|Convention**        |                                                          |
++--------------------+----------------------------------------------------------+
+|ObjectId            |{"$oid": <ObjectId bytes as 23-character, big-endian *hex |
```

Comment:
Is it worth clarifying that it is the individual components of the ObjectID that are big-endian? (see: [DOCS-9787](https://jira.mongodb.org/browse/DOCS-9787)).

Also, should the spec suggest lowercase as a canonical form for hex characters (see: [`oid.json`](https://github.com/mongodb/specifications/blob/master/source/bson-corpus/tests/oid.json))? We can still allow flexibility when decoding.